pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go

```diff
@@ -0,0 +1,191 @@
+package globalroles
+
+import (
+	"fmt"
+
+	"github.com/rancher/rancher/pkg/controllers"
+
+	wrangler "github.com/rancher/wrangler/v2/pkg/name"
+
+	mgmtcontroller "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	rbacv1 "github.com/rancher/wrangler/v2/pkg/generated/controllers/rbac/v1"
+	v1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	localFleetWorkspace            = "fleet-local"
+	fleetWorkspaceClusterRulesName = "fwcr"
+	fleetWorkspaceVerbsName        = "fwv"
+)
+
+// fleetWorkspaceRoleHandler manages ClusterRoles created for the InheritedFleetWorkspacePermissions field. It manages 2 roles:
+// - 1) CR for ResourceRules
+// - 2) CR for WorkspaceVerbs
+type fleetWorkspaceRoleHandler struct {
+	crClient rbacv1.ClusterRoleController
+	crCache  rbacv1.ClusterRoleCache
+	fwCache  mgmtcontroller.FleetWorkspaceCache
+}
+
+func newFleetWorkspaceRoleHandler(management *config.ManagementContext) *fleetWorkspaceRoleHandler {
+	return &fleetWorkspaceRoleHandler{
+		crClient: management.Wrangler.RBAC.ClusterRole(),
+		crCache:  management.Wrangler.RBAC.ClusterRole().Cache(),
+		fwCache:  management.Wrangler.Mgmt.FleetWorkspace().Cache(),
+	}
+}
+
+// ReconcileFleetWorkspacePermissions reconciles backing ClusterRoles created for granting permission to fleet workspaces.
+func (h *fleetWorkspaceRoleHandler) reconcileFleetWorkspacePermissions(gr *v3.GlobalRole) error {
+	if err := h.reconcileResourceRules(gr); err != nil {
+		return fmt.Errorf("error reconciling fleet permissions cluster role: %w", err)
+	}
+	if err := h.reconcileWorkspaceVerbs(gr); err != nil {
+		return fmt.Errorf("error reconciling fleet workspace verbs cluster role: %w", err)
+	}
+
+	return nil
+}
+
+func (h *fleetWorkspaceRoleHandler) reconcileResourceRules(gr *v3.GlobalRole) error {
+	crName := wrangler.SafeConcatName(gr.Name, fleetWorkspaceClusterRulesName)
+	cr, err := h.crCache.Get(crName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if gr.InheritedFleetWorkspacePermissions.ResourceRules == nil {
+		return h.deleteClusterRoleIfNeeded(!apierrors.IsNotFound(err), crName)
+	}
+
+	if apierrors.IsNotFound(err) {
+		_, err := h.crClient.Create(backingResourceRulesClusterRole(gr, crName))
+
+		return err
+	}
```

Comment:
Nit: We check if the error is a not found error 3 times in this section of code.  You could re-write this to minimize the amount of checking that we do:

```suggestion
	if err != nil {
	        if !apierrors.IsNotFound(err){
		        return err
	        }
	        if gr.InheritedFleetWorkspacePermissions.ResourceRules == nil {
	               return h.crClient.Delete(crName, &metav1.DeleteOptions{})
	        }
	        _, err := h.crClient.Create(backingResourceRulesClusterRole(gr, crName))
		return err
	} 
```

Reply:
This suggestion code is not correct. This would always call `Delete` for `GlobalRoles` that doesn't have `InheritedFleetWorkspacePermissions`.

I refactored the code to minimize the not found check.